File: src/my_cpp_pkg/src/action_client.cpp
#include <memory>
#include <string>
#include <sstream>

#include "my_robot_interfaces/action/test.hpp"
#include "rclcpp/rclcpp.hpp"

#include "rclcpp_action/rclcpp_action.hpp"

class TestActionClient : public rclcpp::Node
{
public:
  using Test = my_robot_interfaces::action::Test;
  using GoalHandleTest = rclcpp_action::ClientGoalHandle<Test>;

  explicit TestActionClient(const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions())
  : Node("test_action_client", node_options)
  {
    this->client_ptr_ = rclcpp_action::create_client<Test>(
      this->get_node_base_interface(),
      this->get_node_graph_interface(),
      this->get_node_logging_interface(),
      this->get_node_waitables_interface(),
      "my_action_move");

    this->timer_ = this->create_wall_timer(
      std::chrono::milliseconds(500),
      std::bind(&TestActionClient::send_goal, this));
  }

  void send_goal()
  {
    using namespace std::placeholders;

    this->timer_->cancel();

    if (!this->client_ptr_->wait_for_action_server(std::chrono::seconds(10))) {
      RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
      rclcpp::shutdown();
      return;
    }

    auto goal_msg = Test::Goal();
    goal_msg.secs = 10;

    RCLCPP_INFO(this->get_logger(), "Sending goal");

    auto send_goal_options = rclcpp_action::Client<Test>::SendGoalOptions();
    send_goal_options.goal_response_callback =
      std::bind(&TestActionClient::goal_response_callback, this, _1);
    send_goal_options.feedback_callback =
      std::bind(&TestActionClient::feedback_callback, this, _1, _2);
    send_goal_options.result_callback =
      std::bind(&TestActionClient::result_callback, this, _1);
    this->client_ptr_->async_send_goal(goal_msg, send_goal_options);
  }

private:
  rclcpp_action::Client<Test>::SharedPtr client_ptr_;
  rclcpp::TimerBase::SharedPtr timer_;

  void goal_response_callback(std::shared_future<GoalHandleTest::SharedPtr> future)
  {
    auto goal_handle = future.get();
    if (!goal_handle) {
      RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
    } else {
      RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result");
    }
  }

  void feedback_callback(
    GoalHandleTest::SharedPtr,
    const std::shared_ptr<const Test::Feedback> feedback_msg)
  {
    std::stringstream ss;
    ss << "Next number in sequence received: ";
    for (auto number : feedback_msg->feedback) {
      ss << number << " ";
    }
    RCLCPP_INFO(this->get_logger(), ss.str().c_str());
  }

  void result_callback(const GoalHandleTest::WrappedResult & result)
  {
    switch (result.code) {
      case rclcpp_action::ResultCode::SUCCEEDED:
        break;
      case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
        return;
      case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
        return;
      default:
        RCLCPP_ERROR(this->get_logger(), "Unknown result code");
        return;
    }
    std::stringstream ss;
    ss << "Result received: ";
    for (auto number : result.result->status) {
      ss << number << " ";
    }
    RCLCPP_INFO(this->get_logger(), ss.str().c_str());
    rclcpp::shutdown();
  }
};  // class TestActionClient

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  auto action_server = std::make_shared<TestActionClient>();

  rclcpp::spin(action_server);

  rclcpp::shutdown();
  return 0;
}

File: src/my_py_pkg/my_py_pkg/action_client.py
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node
from functools import partial
from my_robot_interfaces.action import Test

class TestActionClient(Node):

    def __init__(self):
        super().__init__('test_action_client')
        
        self.declare_parameter("secs", 2)
        self.secs_ = self.get_parameter("secs").value
        
        self.get_logger().info('Action Client and ready to request moving robots for {} secs'.format(self.secs_))

        self.action_client_ = None
        self.call_action_server()
    
    def destroy(self):
        self.action_client_.destroy()
        super().destroy_node()

    def call_action_server(self):
        self.action_client_ = ActionClient(self, Test, 'my_action_move')

        while(not self.action_client_.wait_for_server(1.0)):
            self.get_logger().warn("Waiting for server my_action_move to be up")

        goal_request = Test.Goal()
        goal_request.secs = self.secs_

        future = self.action_client_.send_goal_async(goal_request, feedback_callback=self.feedback_callback)

        future.add_done_callback(partial(self.callback_action_performed, secs=goal_request.secs))


    def feedback_callback(self, feedback_msg):
        self.get_logger().info("Feedback received")
        self.get_logger().info("Feedback is '{0}' for action \"my_action_move".format(feedback_msg.feedback))

    def callback_action_performed(self, future, secs):
        try:
            goal_handle = future.result()
            if not goal_handle.accepted:
                self.get_logger().info('Goal rejected :(')
                return

            self.get_logger().info('Goal accepted :)')

            self._get_result_future = goal_handle.get_result_async()
            self._get_result_future.add_done_callback(partial(self.get_result_callback, secs=secs))

        except Exception as e:
            self.get_logger().error("Action call failed %r" % (e,))

    def get_result_callback(self, future, secs):
        try:
            status = future.result().result.status
            self.get_logger().info("Final response is {0} for action \"my_action_move\ with secs={1}".format(status, secs))

            self.destroy()
            # Shutdown after receiving a result
            rclpy.shutdown()
            
        
        except Exception as e:
            self.get_logger().error("Action call response reading failed %r" % (e,))

def main(args=None):
    rclpy.init(args=args)
    test_action_client = TestActionClient()
    rclpy.spin(test_action_client)
    rclpy.shutdown() # last line of any ROS2 .py node

if __name__ == '__main__':
    main()